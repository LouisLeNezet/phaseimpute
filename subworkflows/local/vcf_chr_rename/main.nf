include { BCFTOOLS_ANNOTATE           } from '../../../modules/nf-core/bcftools/annotate/main.nf'
include { BCFTOOLS_INDEX              } from '../../../modules/nf-core/bcftools/index/main.nf'
include { FAITOCHR                    } from '../../../modules/local/faitochr/main.nf'

workflow VCF_CHR_RENAME {
    take:
    ch_vcf          // channel: [ [id], vcf, index ]
    ch_fasta        // channel: [ [id], fasta, fai ]

    main:

    ch_versions = Channel.empty()

    // Generate the chromosome renaming file
    FAITOCHR(ch_fasta.map{ metaG, fasta, fai -> [metaG, fai] })
    ch_versions = ch_versions.mix(FAITOCHR.out.versions)

    // Rename the chromosome without prefix
    BCFTOOLS_ANNOTATE(
        ch_vcf // channel: [ [id], vcf, index ]
            .combine(Channel.of([[],[],[]]))
            .combine(FAITOCHR.out.annot_chr.map{it[1]})
    )
    ch_versions = ch_versions.mix(BCFTOOLS_ANNOTATE.out.versions.first())

    BCFTOOLS_INDEX(BCFTOOLS_ANNOTATE.out.vcf)
    ch_versions = ch_versions.mix(BCFTOOLS_INDEX.out.versions.first())

    ch_vcf_renamed = BCFTOOLS_ANNOTATE.out.vcf
        .combine(BCFTOOLS_INDEX.out.csi, by:0)

    emit:
    vcf_renamed    = ch_vcf_renamed        // [ meta, vcf, csi ]
    versions       = ch_versions           // channel: [ versions.yml ]
}
