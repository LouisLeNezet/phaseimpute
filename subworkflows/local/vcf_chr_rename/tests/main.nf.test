nextflow_workflow {

    name "Test Subworkflow VCF_CHR_RENAME"
    script "../main.nf"

    workflow "VCF_CHR_RENAME"

    tag "subworkflows"
    tag "subworkflows_local"
    tag "subworkflows/vcf_chr_rename"
    tag "vcf_chr_rename"

    tag "bcftools"
    tag "bcftools/annotate"
    tag "bcftools/index"
    tag "faitochr"

    test("Should run without error") {
        config "./nextflow_rename.config"
        when {
            workflow {
                """
                input[0] = Channel.fromList([
                    [
                        [id: "multi"],
                        file("https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/genomics/homo_sapiens/illumina/vcf/NA24385_sv.vcf.gz",checkIfExist:true),
                        file("https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/genomics/homo_sapiens/illumina/vcf/NA24385_sv.vcf.gz.tbi",checkIfExist:true)
                    ],
                    [
                        [id: "chr21"],
                        file("https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/genomics/homo_sapiens/illumina/vcf/chr21/simulated_sv.vcf.gz",checkIfExist:true),
                        file("https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/genomics/homo_sapiens/illumina/vcf/chr21/simulated_sv.vcf.gz.tbi",checkIfExist:true)
                    ]
                ])
                input[1] = Channel.of([
                    [id:"GRCh37"],
                    file("https://raw.githubusercontent.com/nf-core/test-datasets/phaseimpute/data/reference_genome/21_22/hs38DH.chr21_22.fa",checkIfExist:true),
                    file("https://raw.githubusercontent.com/nf-core/test-datasets/phaseimpute/data/reference_genome/21_22/hs38DH.chr21_22.fa.fai",checkIfExist:true)
                ])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match() }
            )
        }
    }
    test("Should run with error") {
        config "./nextflow.config"
        when {
            workflow {
                """
                input[0] = Channel.fromList([
                    [
                        [id: "chr22"],
                        file("https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/genomics/homo_sapiens/illumina/vcf/test.rnaseq.vcf.gz",checkIfExist:true),
                        file("https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/genomics/homo_sapiens/illumina/vcf/test.rnaseq.vcf.gz.tbi",checkIfExist:true)
                    ],
                    [
                        [id: "chr21"],
                        file("https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/genomics/homo_sapiens/illumina/vcf/chr21/simulated_sv.vcf.gz",checkIfExist:true),
                        file("https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/genomics/homo_sapiens/illumina/vcf/chr21/simulated_sv.vcf.gz.tbi",checkIfExist:true)
                    ]
                ])
                input[1] = Channel.of([
                    [id:"GRCh37"],
                    file("https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/genomics/homo_sapiens/genome/genome.fasta",checkIfExist:true),
                    file("https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/genomics/homo_sapiens/genome/genome.fasta.fai",checkIfExist:true)
                ])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.failed },
            )
        }
    }
}
