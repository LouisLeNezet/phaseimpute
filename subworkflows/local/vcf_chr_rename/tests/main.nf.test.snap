{
    "Should run without error": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "chr21"
                        },
                        "chr21_chrrename.vcf.gz:md5,39cd8e316cd9b9282b8289d69d81260b",
                        "chr21_chrrename.vcf.gz.csi:md5,3bbbb50b0dd3515d380eabe0013cde19"
                    ],
                    [
                        {
                            "id": "multi"
                        },
                        "multi_chrrename.vcf.gz:md5,5f6f1ca261270d55eec054368f3d9587",
                        "multi_chrrename.vcf.gz.csi:md5,5d175780d5611d962430bff3377f649f"
                    ]
                ],
                "1": [
                    "versions.yml:md5,176431a832f84d4c329f6d1e9c74d203",
                    "versions.yml:md5,3698013e288e15d392e1cd3e22d2022a",
                    "versions.yml:md5,924dd2e49d998f8f0da93799e62196f7"
                ],
                "vcf_renamed": [
                    [
                        {
                            "id": "chr21"
                        },
                        "chr21_chrrename.vcf.gz:md5,39cd8e316cd9b9282b8289d69d81260b",
                        "chr21_chrrename.vcf.gz.csi:md5,3bbbb50b0dd3515d380eabe0013cde19"
                    ],
                    [
                        {
                            "id": "multi"
                        },
                        "multi_chrrename.vcf.gz:md5,5f6f1ca261270d55eec054368f3d9587",
                        "multi_chrrename.vcf.gz.csi:md5,5d175780d5611d962430bff3377f649f"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,176431a832f84d4c329f6d1e9c74d203",
                    "versions.yml:md5,3698013e288e15d392e1cd3e22d2022a",
                    "versions.yml:md5,924dd2e49d998f8f0da93799e62196f7"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-03-24T11:45:51.636035695"
    }
}